### DISCUSSION TITLE
ADB Forward to devices in a farm

### DISCUSSION THREAD

**Discussion by ConfusedMerlin** [DeviceFarmer/stf#579] [Q&A] [answered]
Greetings,

I want to ask if it should be possible to forward and/or reverse ports to a device connected to a devicefarm., using "adb connect server:port" as shown at the device's detail page. 

Setup is my dev-linux, where at localhost:8080 a http server is sitting. I want that 8080 to be accessible at the device in that farm at 8081, so I do "adb reverse tcp:8081 tcp:8080". That call returns "8080" , and on the shell of the device inside the farm I can see a (v6) localhost 8081 with "netstat -atn".

So far, so good. But when opening the chrome browser on that device and entering "localhost.8081", it says that localhost did not return anything. And the forward port does not disappear when the adb remote connection is closed. Also, netcat that port on localhost on the device's shell will return nothing (it should stick in a stdin). 

So the question is, should that be possible at all?

**Comment by denis99999** [comment, marked as answer]
@ConfusedMerlin , => https://github.com/devicefarmer/adbkit#devicereverseremote-local

So as the reverse port forwarding is created on the remote ADB server, you can redirect that flow to your considered http server, for that you can use the `socat` command for instance on the remote ADB server, something like the following should work: 
`# socat TCP4-LISTEN:8080,reuseaddr,fork TCP4:YOUR_dev-linux_IP:8080`

**Reply by ConfusedMerlin** [reply]
> @ConfusedMerlin , => https://github.com/devicefarmer/adbkit#devicereverseremote-local
> 
> So as the reverse port forwarding is created on the remote ADB server, you can redirect that flow to your considered http server, for that you can use the `socat` command for instance on the remote ADB server, something like the following should work: `# socat TCP4-LISTEN:8080,reuseaddr,fork TCP4:YOUR_dev-linux_IP:8080`

So the "adb reverse" from the dev-linux causes the devicefarm-adb inside the farm to "reverse" the port 8080 from the devicefarm-device to the devicefarm-adb, but the adb on the dev-linux gets nothing?

**Reply by denis99999** [reply]
> So the "adb reverse" from the dev-linux causes the devicefarm-adb inside the farm to "reverse" the port 8080 from the devicefarm-device to the devicefarm-adb, but the adb on the dev-linux gets nothing?

Yes as it is specified in `adbkit` => `Reverses socket connections from the device (remote) to the ADB server host (local). This is analogous to adb reverse <remote> <local>. It's important to note that if you are connected to a remote ADB server, the reverse will be created on that host.`

**Reply by ConfusedMerlin** [reply]
> > So the "adb reverse" from the dev-linux causes the devicefarm-adb inside the farm to "reverse" the port 8080 from the devicefarm-device to the devicefarm-adb, but the adb on the dev-linux gets nothing?
> 
> Yes as it is specified in `adbkit` => `Reverses socket connections from the device (remote) to the ADB server host (local). This is analogous to adb reverse <remote> <local>. It's important to note that if you are connected to a remote ADB server, the reverse will be created on that host.`

I see....  thank you for pointing this out. I must ponder about how we can make this work in a very structured network without to many extra firewall-rules.

**Comment by denis99999** [comment]
Another solution is simply to use the 'Advanced->Port forwarding' feature at UI level, it should fit your need, for instance : `Device:8080 => dev-linux:8080`
